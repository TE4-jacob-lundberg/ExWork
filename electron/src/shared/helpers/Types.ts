export interface ILink {
  url: string;
  label: string;
}

export interface IGameFormFields {
  id: string;
  title: string;
  abbreviation: string;
  image: string; 
  xAxis: string;
  yAxis: string;
  links: ILink[];
  fileNames: string[];
}
