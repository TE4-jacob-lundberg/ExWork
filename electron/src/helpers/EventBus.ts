/* eslint-disable @typescript-eslint/no-explicit-any */
const EventBus: IEventBus = {
  events: {},
  dispatch: function (event: string, data: any): void {
    if (!this.events[event]) return;
    this.events[event].forEach((callBack: Function) => callBack(data));
  },
  subscribe: function (event: string, callBack: Function): void {
    if (!this.events[event]) this.events[event] = [];
    this.events[event].push(callBack);
  },
};

interface IEventBus {
  events: {[id: string]: Function[]};
  dispatch: (event: string, data: any) => void;
  subscribe: (event: string, callBack: Function) => void;
}

export { EventBus };
