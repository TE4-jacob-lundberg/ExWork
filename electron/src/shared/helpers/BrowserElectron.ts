// eslint-disable-next-line @typescript-eslint/no-explicit-any
export let electron: any;

if (process.env.ENVIRONMENT === 'test') electron = require('electron');
else electron = window.require('electron');
