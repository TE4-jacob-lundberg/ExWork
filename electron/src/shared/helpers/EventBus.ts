/* eslint-disable @typescript-eslint/no-explicit-any */
interface IEventBus {
  events: {[id: string]: Function[]};
  dispatch: (event: string, data: any) => void;
  subscribe: (event: string, callBack: Function) => void;
}

const EventBus: IEventBus = {
  events: {},
  dispatch: function (event: string, data: any): void {
    if (!this.events[event]) return;
    this.events[event].forEach((callBack: Function) => callBack(data));
  },
  subscribe: function (event: string, callBack: Function): void {
    if (!this.events[event]) this.events[event] = [];
    this.events[event].push(callBack);
  },
};

export { EventBus };
