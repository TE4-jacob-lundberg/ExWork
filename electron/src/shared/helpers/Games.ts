import { standardGames, IGame } from '../constants/standardGames';

export class Games {

  public static all(): IGame[] {
    if (!localStorage.getItem('games')) localStorage.setItem('games', JSON.stringify(standardGames));

    return JSON.parse(localStorage.getItem('games')!) || [];
  }

  public static addNew(game: IGame): string {
    const savedGames: IGame[] = JSON.parse(localStorage.getItem('games')!);
    if (savedGames.some(sGame => sGame.id === game.id)) return 'DUPLICATE_ID';
    
    localStorage.setItem('games', JSON.stringify(savedGames.concat(game)));
    return '';
  }
}
