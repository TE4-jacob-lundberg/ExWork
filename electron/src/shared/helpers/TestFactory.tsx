/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';
import { render, RenderResult } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { createBrowserHistory } from 'history';

export function factory(component: JSX.Element): RenderResult {
  return render(component);
}

export function routerFactory(component: JSX.Element, options: any = { history: createBrowserHistory() }): RenderResult {
  return factory(
    <Router history={options.history}>
      {component}
    </Router>
  );
}
