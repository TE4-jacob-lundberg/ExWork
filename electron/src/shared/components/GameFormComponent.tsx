import React, { useState } from 'react';
import { useFormik } from 'formik';
import Styled from '@emotion/styled';
import { css } from '@emotion/core';

import { IGameFormFields, IGameFormErrors } from '../helpers/Types';
import ButtonComponent from '../components/ButtonComponent';
import ModalComponent from '../components/ModalComponent';

const ContainerStyled = Styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr;
  height: 100%;
  width: 100%;
  position: relative;
  border-radius: 5px;
  
  & > form {
    height: 100%;
    padding: 16px;
    background: #333;
    width: 100%;
  }
`;

const SectionHeaderStyled = Styled.h1`
  width: fit-content;
  margin: 12px 0;
`;

const SectionStyled = Styled.section`
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-gap: 16px;
  padding-bottom: 20px;
  border-bottom: solid 1px #555;


  & > input {
    width: auto;
  }
`;

type LabelProps = {
  error?: boolean;
}

const InputLabelStyled = Styled.label<LabelProps>`
  color: ${(props): string => props.error ? 'red' : ''};
  display: flex;
  align-items: center;
`;

const InputStyled = Styled.input`
  height: 24px;
  &:focus {
    outline: none;
  }
`;

type RangeProps = {
  value: string;
}

const RangeInputStyled = Styled.input<RangeProps>`
  position: relative;
  width: 100%;
  &::after {
    content: '${(props): string => props.value}';
    position: absolute;
    left: calc(${(props): string => (Number(props.value)).toString()}% - 8px);
    top: -12px;
  } 
`;

type ImageProps = {
  image: string;
  x: string;
  y: string;
}

const ImageStyled = Styled.div<ImageProps>`
  width: 23.33vw;
  margin: auto;
  height: 100%;
  background: url(${(props): string => `${props.image}) ${props.x}% ${props.y}%`} no-repeat;
  background-size: cover;
`;

const ErrorMessageStyled = Styled.li`
  color: red;
  margin: 12px 0;
`;

type ActionsContainerProps = {
  deletable: boolean;
}

const ActionsContainer = Styled.div<ActionsContainerProps>`
  position: absolute;
  bottom: 12px;
  left: 12px;
  display: flex;
  flex-direction: ${(props): string => props.deletable ? 'row' : 'row-reverse'};
  justify-content: space-between;
  width: calc(70vw / 2);
  padding: 0 16px;

  & > button {
    margin: 0 4px;
  }
`;

interface Props {
  initialValues?: IGameFormFields;
  deletable?: boolean;
  onSubmit: (values: IGameFormFields) => void;
  onDelete?: () => void;
}

const GameFormComponent: React.FC<Props> = function (props: Props) {
  const [showModal, setShowModal] = useState(false);
  const formik = useFormik({
    initialValues: props.initialValues || {title: '', abbreviation: '', image: '', xAxis: '50', yAxis: '0', fileNames: [localStorage.getItem('latestGameFile')!]},
    onSubmit: (values): void => {
      props.onSubmit(values);
    },
    validate: (values): IGameFormErrors => {
      const errors: IGameFormErrors = {};

      if (!values.title) errors.title = 'You need to enter a title for the game';
      if (!values.image) errors.image = 'You need to select an image to represent the game';

      if (Object.values(errors).find(Boolean)) setShowModal(true);

      return errors;
    },
    validateOnBlur: false,
    validateOnChange: false,
  });

  function handleImageChange(e: React.ChangeEvent<HTMLInputElement>): void {
    const reader = new FileReader();

    reader.addEventListener('load', () => {
      formik.setValues({...formik.values, image: reader.result as string });
    });

    reader.readAsDataURL(e.target.files![0]);
  }

  return (
    <>
      <ContainerStyled>
        <form onSubmit={formik.handleSubmit}>
          <SectionHeaderStyled>Game Information</SectionHeaderStyled>
          <SectionStyled>
            <InputLabelStyled htmlFor="title" error={!!formik.errors.title}>Title*</InputLabelStyled>
            <InputStyled type="text" {...formik.getFieldProps('title')} id="title" />          
            <InputLabelStyled htmlFor="abbr">Abbreviation</InputLabelStyled>
            <InputStyled type="text" {...formik.getFieldProps('abbreviation')} id="abbr" />
            <InputLabelStyled>Game files (*.exe, *.app)</InputLabelStyled>
            <InputStyled type="text" {...formik.getFieldProps('fileNames')} />
          </SectionStyled>
          <SectionHeaderStyled>Image Settings</SectionHeaderStyled>
          <SectionStyled>
            <InputLabelStyled error={!!formik.errors.image}>Image*</InputLabelStyled>
            <InputStyled type="file" 
              onChange={handleImageChange} 
            />
            {formik.values.image && (
              <>
                <InputLabelStyled>X-axis</InputLabelStyled>
                <RangeInputStyled type="range" min="0" max="100" {...formik.getFieldProps('xAxis')} />
                <InputLabelStyled>Y-axis</InputLabelStyled>
                <RangeInputStyled type="range" min="0" max="100" {...formik.getFieldProps('yAxis')} />
              </>)}
          </SectionStyled>
          <ActionsContainer deletable={!!props.deletable}>
            {props.deletable && <ButtonComponent
              onClick={(): void => props.onDelete && props.onDelete()}
              background="red"
            >
              Delete 
              <i className="material-icons">delete</i>
            </ButtonComponent>}
            <ButtonComponent 
              onClick={(): string => ''}
              type="submit"
              background="green"
              styling={css`            
              & > i {
                margin-left: 8px;
              }
              `}
              data-testid="submit-form"
            >
              Save
              <i className="material-icons">save</i>
            </ButtonComponent>
          </ActionsContainer>
        </form>
        <ImageStyled 
          image={formik.values.image} 
          x={formik.values.xAxis}
          y={formik.values.yAxis}
        />
      </ContainerStyled>
      {showModal && <ModalComponent
        title="You need to fix these errors to save your game"
        onClose={(): void => setShowModal(false)}
        onContinue={(): void => setShowModal(false)}
      >
        <ul>
          {Object.values(formik.errors).filter(Boolean).map((value, i) => (<ErrorMessageStyled key={i}>{value || ''}</ErrorMessageStyled>))}
        </ul>
      </ModalComponent>}
    </>
  );
};

export default GameFormComponent;
