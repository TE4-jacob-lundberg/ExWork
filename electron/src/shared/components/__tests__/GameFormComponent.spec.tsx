import React from 'react';
import { fireEvent, waitForElement } from '@testing-library/react';

import { factory } from '../../helpers/TestFactory';
import GameFormComponent from '../GameFormComponent';

const mockedSubmit = jest.fn();

const testComponent = <GameFormComponent 
  onSubmit={mockedSubmit}
/>;

describe('GameFormComponent.tsx', () => {
  it('renders correctly', () => {
    const { container } = factory(testComponent);
    
    expect(container).toMatchSnapshot();
  });

  it('submits with the correct values', () => {
    const { getByTestId } = factory(testComponent);

    fireEvent.change(getByTestId('game-id'), { target: { value: 'test' }});
    fireEvent.change(getByTestId('game-title'), { target: { value: 'test-title' }});
    fireEvent.change(getByTestId('game-abbr'), { target: { value: 'test-abbr' }});
    
    fireEvent.click(getByTestId('submit-form'));

    waitForElement(() => getByTestId('game-abbr')).then(() => {
      expect(mockedSubmit).toHaveBeenCalledWith({
        id: 'test', title: 'test-title', abbreviation: 'test-abbr', image: '', xAxis: '50', yAxis: '0', links: [], fileNames: [],
      });
    });
  });
});
