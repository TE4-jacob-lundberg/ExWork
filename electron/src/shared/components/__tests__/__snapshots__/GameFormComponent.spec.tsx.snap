// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameFormComponent.tsx renders correctly 1`] = `
<div>
  <div
    class="css-1oynk9y"
  >
    <form>
      <h1
        class="css-1vjmso2"
      >
        Game Information
      </h1>
      <section
        class="css-6a1pbd"
      >
        <label
          class="css-13m3dri"
          for="title"
        >
          Title*
        </label>
        <input
          class="css-1o0il4p"
          id="title"
          name="title"
          type="text"
          value=""
        />
        <label
          class="css-13m3dri"
          for="abbr"
        >
          Abbreviation
        </label>
        <input
          class="css-1o0il4p"
          id="abbr"
          name="abbreviation"
          type="text"
          value=""
        />
        <label
          class="css-13m3dri"
        >
          Game files (*.exe, *.app)
        </label>
        <input
          class="css-1o0il4p"
          name="fileNames"
          type="text"
          value=""
        />
      </section>
      <h1
        class="css-1vjmso2"
      >
        Image Settings
      </h1>
      <section
        class="css-6a1pbd"
      >
        <label
          class="css-13m3dri"
        >
          Image*
        </label>
        <input
          class="css-1o0il4p"
          type="file"
        />
        
      </section>
      <div
        class="css-1qnhe9y"
      >
        <button
          class="css-1ctwf7v"
          data-testid="submit-form"
          type="submit"
        >
          Save
          <i
            class="material-icons"
          >
            save
          </i>
        </button>
      </div>
    </form>
    <div
      class="css-1nso4la"
      x="50"
      y="0"
    />
  </div>
</div>
`;
